File: typescript/src/day4.ts
import { readFileSync } from "fs"

const input = readFileSync("../inputs/day4.txt", "utf-8")

// directional steps are horizontal, vertical and diagonal
const DIRECTIONS: [number, number][] = [
    // Horizontal
    [1, 0],
    [-1, 0],
    // Vertical
    [0, 1],
    [0, -1],
    // Diagonal
    [1, 1],
    [1, -1],
    [-1, -1],
    [-1, 1],
]

export const isNotExpectedLetter = (
    grid: string[][],
    expectedLetter: string,
    col: number,
    row: number
) => {
    // This is lazy. I don't want to deal with it if we go outside the bounds of the grid...
    try {
        return grid[row][col] !== expectedLetter
    } catch {
        return true
    }
}

export const part1 = (input: string) => {
    const grid: string[][] = input.split("\r\n").map((line) => line.split(""))

    const wordToFind: string = "XMAS"

    let result: number = 0

    // Iterate through every item in the grid
    for (let row = 0; row < grid.length; row++) {
        for (let col = 0; col < grid[row].length; col++) {
            DIRECTIONS.forEach((direction) => {
                let foundXmas: boolean = true

                for (let index = 0; index < wordToFind.length; index++) {
                    if (
                        isNotExpectedLetter(
                            grid,
                            wordToFind[index],
                            // if X is at index 1, and it's horizontal forwards, we expect
                            // That M (index 1 in the word) will be at (0, 1) from where we are
                            // A (index 2) at (0, 2) from where we are or 2 x the direction value
                            // and S (index 3) at (0, 3) or 3x direction values
                            // This applies diagonally, backwards, wherever we like
                            col + index * direction[0],
                            row + index * direction[1]
                        )
                    ) {
                        // Stop checking letters in the word if we don't find the letter we expect
                        foundXmas = false
                        break
                    }
                }

                if (foundXmas) result++
            })
        }
    }

    return result
}

const part1Result = part1(input)

console.log("The result for Day 4 part 1 is: ", part1Result)

File: rust/src/bin/day4.rs
use std::collections::HashSet;

const INPUT: &str = include_str!("../../../inputs/day4.txt");

fn main() {
    println!("The result for day 4 part 1 is {}", part1(INPUT));
    println!("The result for day 4 part 2 is {}", part2(INPUT));
}

const DIRECTIONS: [[i32; 2]; 8] = [
    // Horizontal
    [1, 0],
    [-1, 0],
    // Vertical
    [0, 1],
    [0, -1],
    // Diagonal
    [1, 1],
    [1, -1],
    [-1, -1],
    [-1, 1],
];

fn is_not_xmas(grid: &Vec<Vec<char>>, expected_letter: char, col: usize, row: usize) -> bool {
    if row >= grid.len() || col >= grid[row].len() {
        return true; // Out of bounds
    }
    grid[row][col] != expected_letter
}

fn part1(input: &str) -> i32 {
    let mut result: i32 = 0;
    let word_to_find: Vec<char> = vec!['X', 'M', 'A', 'S'];

    // Iterate over each line, and collect each char in that line into a vector
    //  which gets pushed to an overall vector of vectors of chars (a grid)
    let grid = input
        .lines()
        .map(|line| line.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();

    // Iterate through each char in the grid and check if there's an X there
    for row in 0..grid.len() {
        for col in 0..grid[row].len() {
            // Iterate through the directions vector
            for &[x, y] in DIRECTIONS.iter() {
                let mut found_xmas = true;

                // Loop through the chars in "XMAS", start at X and look for each one in turn, in every direction
                for (index, &letter) in word_to_find.iter().enumerate() {
                    let new_col = col as i32 + index as i32 * x;
                    let new_row = row as i32 + index as i32 * y;

                    // Check boundaries, if we're outside the boundaries or the letter doesn't match XMAS then set false
                    if new_row < 0
                        || new_col < 0
                        || new_row as usize >= grid.len()
                        || new_col as usize >= grid[new_row as usize].len()
                        || is_not_xmas(&grid, letter, new_col as usize, new_row as usize)
                    {
                        found_xmas = false;
                        break;
                    }
                }
                // Otherwise we've found XMAS and we can increment the value
                if found_xmas {
                    result += 1;
                }
            }
        }
    }

    result
}

fn part2(input: &str) -> i32 {
    let mut result: i32 = 0;

    // Input to a grid, let's turn this into a utility function later
    let grid = input
        .lines()
        .map(|line| line.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();

    // Iterate through the grid looking for "A"
    for row in 0..grid.len() {
        for col in 0..grid[row].len() {
            if grid[row][col] == 'A' {
                // Check both diagonal directions

                let mut forward: HashSet<char> = HashSet::new();
                let mut backward: HashSet<char> = HashSet::new();

                // Check we're not out of bounds and add the letter if it's in bounds
                if row > 0 && col > 0 {
                    forward.insert(grid[row - 1][col - 1]);
                }

                if row + 1 < grid.len() && col + 1 < grid[row + 1].len() {
                    forward.insert(grid[row + 1][col + 1]);
                }

                if row > 0 && col + 1 < grid[row].len() {
                    backward.insert(grid[row - 1][col + 1]);
                }

                if row + 1 < grid.len() && col > 0 {
                    backward.insert(grid[row + 1][col - 1]);
                }

                // Now we just care if each set has an S and an M
                // Because there are two elements in each set
                // we just need to check for each letter
                if forward.contains(&'S')
                    && forward.contains(&'M')
                    && backward.contains(&'S')
                    && backward.contains(&'M')
                {
                    result += 1;
                }
            }
        }
    }

    result
}

#[cfg(test)]
mod tests {
    const TEST_INPUT: &str = include_str!("../../../testInputs/day4.txt");
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(part1(TEST_INPUT), 18);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(TEST_INPUT), 9);
    }
}
